{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ec8427d-1172-4f55-b9b7-dc266ad274d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.SeqUtils import gc_fraction \n",
    "import numpy as np\n",
    "import subprocess\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb2df944-372a-4c11-a6bd-2c9a6787f159",
   "metadata": {},
   "outputs": [],
   "source": [
    "fp = \"/home/myoui/shared/ML/Homo_sapiens.GRCh38.ncrna.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "004ca6aa-87db-4f07-9495-b12eda22fb59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Column 1 list length: 203776\n",
      "Column 2 list length: 203776\n",
      "Column 3 list length: 203776\n",
      "Column 4 list length: 203776\n",
      "Column 5 list length: 203776\n",
      "Column 6 list length: 203776\n",
      "Column 7 list length: 203776\n",
      "Column 8 list length: 203776\n",
      "Sequence list length: 203776\n"
     ]
    }
   ],
   "source": [
    "data_lists = [[] for _ in range(8)]  # Creates a total of 8 lists for columns one to eight\n",
    "sequence_list = []\n",
    "with open(fp, 'r') as file:\n",
    "    current_sequence = []  \n",
    "    for line in file:\n",
    "        line = line.strip()  # Removes whitespace\n",
    "\n",
    "        # Processes header lines starting with '>'\n",
    "        if line.startswith('>'):\n",
    "            # Before processing the new header, appends the current sequence if it exists\n",
    "            if current_sequence:\n",
    "                sequence_list.append(''.join(current_sequence))  # Joins and appends sequence\n",
    "                current_sequence = []  # Resets for the next sequence\n",
    "\n",
    "            # Splits line into fields\n",
    "            parts = line.split()\n",
    "\n",
    "            # Appends to respective lists only if there are at least 8 parts\n",
    "            for i in range(8):\n",
    "                if i < len(parts):\n",
    "                    data_lists[i].append(parts[i].replace('>', '') if i == 0 else parts[i])\n",
    "                else:\n",
    "                    data_lists[i].append(None)  # Fills with None if the part doesn't exist\n",
    "\n",
    "        else:\n",
    "            # If it's a sequence line (not a header), accumulate the sequence\n",
    "            current_sequence.append(line)\n",
    "\n",
    "    # Append the last sequence after the loop\n",
    "    if current_sequence:\n",
    "        sequence_list.append(''.join(current_sequence))\n",
    "\n",
    "# Prints lengths to check for mismatches\n",
    "for idx, lst in enumerate(data_lists):\n",
    "    print(f\"Column {idx + 1} list length: {len(lst)}\")\n",
    "print(f\"Sequence list length: {len(sequence_list)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e36217a0-e73f-479e-8ec9-abc676c9f454",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the parsed data in dataframe\n",
    "df = pd.DataFrame(data_lists).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "317bda34-2f40-458c-b19e-93d88e6c1106",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Columns for Gene list dataframe\n",
    "transcript_id = df[0]\n",
    "chromosome_location = df[2]\n",
    "ensembl_id = df[3]\n",
    "gene_type = df[4]\n",
    "gene_symbol = df[6]\n",
    "\n",
    "# Removes prefixes\n",
    "ensembl_id = ensembl_id.str.replace('gene:', '', regex=False)\n",
    "gene_type = gene_type.str.replace('gene_biotype:', '', regex=False)\n",
    "gene_symbol = gene_symbol.str.replace('gene_symbol:', '', regex=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1114be7b-cd26-42db-b3fc-5aec528f629e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Gene list dataframe\n",
    "gene_list = pd.DataFrame({\n",
    "    'transcript_id' : transcript_id,\n",
    "    'ensembl_id': ensembl_id,\n",
    "    'chromosome_location' : chromosome_location,\n",
    "    'gene_type' : gene_type,\n",
    "    'gene_symbol' : gene_symbol,\n",
    "    'sequence' : sequence_list\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6dd01756-9126-43d0-ace7-508aa954e5b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the DataFrame to CSV\n",
    "output = '/home/myoui/shared/ML/Gene_list.csv'\n",
    "gene_list.to_csv(output, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6eb5a2aa-6715-49fa-8b95-010e853803e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['miRNA', 'Mt_tRNA', 'vault_RNA', 'rRNA', 'sRNA', 'scRNA',\n",
       "       'Mt_rRNA', 'snRNA', 'scaRNA', 'snoRNA', 'ribozyme', 'lncRNA',\n",
       "       'misc_RNA'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list['gene_type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3256e76-f6d7-454c-95f3-bd77fb2112e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_type\n",
       "lncRNA       196132\n",
       "misc_RNA       2419\n",
       "snRNA          2094\n",
       "miRNA          1945\n",
       "snoRNA         1020\n",
       "rRNA             71\n",
       "scaRNA           51\n",
       "Mt_tRNA          22\n",
       "ribozyme          9\n",
       "sRNA              6\n",
       "vault_RNA         4\n",
       "Mt_rRNA           2\n",
       "scRNA             1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list['gene_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "edb58a20-ec58-41b6-87cf-8e43bd891986",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Categories with significant no. of genes\n",
    "categories = ['lncRNA', 'misc_RNA', 'snRNA', 'miRNA', 'snoRNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7aeff706-c34a-4cf6-ab29-e391d2ffcf41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter only gene categories except lncRNA\n",
    "gene_filtered = gene_list[gene_list['gene_type'].isin(categories) & (gene_list['gene_type'] != 'lncRNA')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed9d2cb3-df96-46b9-91e6-0d74bd348ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomly select 2000 genes with random_state = 42 for reproducibility\n",
    "lncrna_sample = gene_list[gene_list['gene_type'] == 'lncRNA'].sample(n=2000, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "55495902-be19-4c46-8472-d1b66cd3d1c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add lncRNA\n",
    "gene_filtered = pd.concat([gene_filtered, lncrna_sample])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f1493317-f212-43e6-b6a8-1593f8c6445a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get GC content and Sequence length\n",
    "def gc_content(seq):\n",
    "    return gc_fraction(seq)\n",
    "\n",
    "def sequence_length(seq):\n",
    "    return len(seq)\n",
    "\n",
    "gene_filtered['gc_content'] = gene_filtered['sequence'].apply(gc_content)\n",
    "gene_filtered['seq_length'] = gene_filtered['sequence'].apply(sequence_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0b5c44dd-5b87-4b0d-9b4e-e18d769e923c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>chromosome_location</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>seq_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105314</th>\n",
       "      <td>ENST00000717492.1</td>\n",
       "      <td>ENSG00000245694.13</td>\n",
       "      <td>chromosome:GRCh38:16:54905876:54930578:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>CRNDE</td>\n",
       "      <td>AGACTCTCGAGCGCCGGCCCCAGGATGACAATCACATCCCAGGAGC...</td>\n",
       "      <td>0.409809</td>\n",
       "      <td>19697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123700</th>\n",
       "      <td>ENST00000663612.1</td>\n",
       "      <td>ENSG00000224078.16</td>\n",
       "      <td>chromosome:GRCh38:15:25094037:25122467:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>SNHG14</td>\n",
       "      <td>GGATCGATGATGACTTTTATACATGCATTCCTTGGAAAGCTGAACA...</td>\n",
       "      <td>0.388336</td>\n",
       "      <td>7339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11323</th>\n",
       "      <td>ENST00000659277.1</td>\n",
       "      <td>ENSG00000224924.10</td>\n",
       "      <td>chromosome:GRCh38:21:20735784:20803297:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LINC00320</td>\n",
       "      <td>AATCATCTCCCATTGAAAGCTGAATCATGAGGATACCATCTCCAAG...</td>\n",
       "      <td>0.395783</td>\n",
       "      <td>4932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100329</th>\n",
       "      <td>ENST00000522627.1</td>\n",
       "      <td>ENSG00000253311.4</td>\n",
       "      <td>chromosome:GRCh38:5:159776772:159871384:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>LINC01847</td>\n",
       "      <td>GTTGCATTAAGTCATCGAGTTTGTGGTGCTTTGTTATGGGAGCCTT...</td>\n",
       "      <td>0.414304</td>\n",
       "      <td>5607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38960</th>\n",
       "      <td>ENST00000443641.5</td>\n",
       "      <td>ENSG00000234893.6</td>\n",
       "      <td>scaffold:GRCh38:HSCHR6_MHC_QBL_CTG1:1546099:15...</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>HCG18</td>\n",
       "      <td>CAATACTCGCCGGAAGCGGAAGCCGCGCCGAGGCTCGTGTAAAAGT...</td>\n",
       "      <td>0.440120</td>\n",
       "      <td>4317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18874</th>\n",
       "      <td>ENST00000415033.4</td>\n",
       "      <td>ENSG00000235903.12</td>\n",
       "      <td>chromosome:GRCh38:13:46052815:46116576:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>CPB2-AS1</td>\n",
       "      <td>CTTCGTGGACACGCCCTCTAGGCGTAAGGTTCCATGCGAGCGGTAA...</td>\n",
       "      <td>0.392840</td>\n",
       "      <td>4078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50616</th>\n",
       "      <td>ENST00000452075.7</td>\n",
       "      <td>ENSG00000293313.2</td>\n",
       "      <td>chromosome:GRCh38:10:42513510:42552794:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ZNF37BP</td>\n",
       "      <td>GTTTTTCTCCGGTACAGCCTGGTGAAGGATGCCCGGAGGCCTCGGC...</td>\n",
       "      <td>0.403728</td>\n",
       "      <td>8637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132712</th>\n",
       "      <td>ENST00000624299.1</td>\n",
       "      <td>ENSG00000247809.12</td>\n",
       "      <td>chromosome:GRCh38:15:96285068:96289265:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>NR2F2-AS1</td>\n",
       "      <td>TTTTCTAATAATTTTTTTTTTTTTTTTTTGAGACAGAGTCTCGCTC...</td>\n",
       "      <td>0.445688</td>\n",
       "      <td>4198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119139</th>\n",
       "      <td>ENST00000668181.1</td>\n",
       "      <td>ENSG00000259234.9</td>\n",
       "      <td>chromosome:GRCh38:15:79187239:79283949:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>ANKRD34C-AS1</td>\n",
       "      <td>AGTTAGAAGCAGATCAGAAATCGAACACGTTTGCTTTTGCGATTCG...</td>\n",
       "      <td>0.491772</td>\n",
       "      <td>5165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105431</th>\n",
       "      <td>ENST00000711040.1</td>\n",
       "      <td>ENSG00000250337.11</td>\n",
       "      <td>chromosome:GRCh38:5:27472227:27488861:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>PURPL</td>\n",
       "      <td>GGAGCCATGAGGAATTCATGCCTTGCAAAGGGGAGGAGCCTGCCCT...</td>\n",
       "      <td>0.342486</td>\n",
       "      <td>4625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185086</th>\n",
       "      <td>ENST00000653715.1</td>\n",
       "      <td>ENSG00000230461.11</td>\n",
       "      <td>chromosome:GRCh38:1:213817996:213966287:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>PROX1-AS1</td>\n",
       "      <td>CAGAGAGAACCATTCTCTGCTTTCAATAGGACAGAAGAGCTCCTTG...</td>\n",
       "      <td>0.453307</td>\n",
       "      <td>5140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67677</th>\n",
       "      <td>ENST00000484322.5</td>\n",
       "      <td>ENSG00000242593.9</td>\n",
       "      <td>chromosome:GRCh38:7:124032205:124395118:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>description:novel</td>\n",
       "      <td>GCCCCAGTACGTTGAAGAAGAAGGTGAAGAAGATCTCATCCAGAAA...</td>\n",
       "      <td>0.357077</td>\n",
       "      <td>4310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106417</th>\n",
       "      <td>ENST00000504820.2</td>\n",
       "      <td>ENSG00000249494.7</td>\n",
       "      <td>chromosome:GRCh38:5:119007965:119070902:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>DMXL1-DT</td>\n",
       "      <td>AATCTCTTTGAGGCTTCTACAGGAGAACGCCGTGCGTTTGAGCTTG...</td>\n",
       "      <td>0.410514</td>\n",
       "      <td>4375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48835</th>\n",
       "      <td>ENST00000650366.1</td>\n",
       "      <td>ENSG00000229807.14</td>\n",
       "      <td>chromosome:GRCh38:X:73820684:73846270:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>XIST</td>\n",
       "      <td>ACACAGCTCAACCTATCTGAATAAAATCCTACTCTCAGACCCCTTT...</td>\n",
       "      <td>0.399124</td>\n",
       "      <td>12555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150406</th>\n",
       "      <td>ENST00000618132.2</td>\n",
       "      <td>ENSG00000251562.12</td>\n",
       "      <td>chromosome:GRCh38:11:65501009:65506110:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>TTTTTTTTTTTTACACGAATTTGAGGAAAACCAAATGAATTTGATA...</td>\n",
       "      <td>0.387116</td>\n",
       "      <td>4983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150518</th>\n",
       "      <td>ENST00000657283.1</td>\n",
       "      <td>ENSG00000254092.5</td>\n",
       "      <td>chromosome:GRCh38:8:20950392:21112728:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>description:novel</td>\n",
       "      <td>TTCACCATGTTGGTCAGGCTGGTCTCGAATGCCTGACCTCAGGTGA...</td>\n",
       "      <td>0.402561</td>\n",
       "      <td>4685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156796</th>\n",
       "      <td>ENST00000673654.1</td>\n",
       "      <td>ENSG00000172965.18</td>\n",
       "      <td>chromosome:GRCh38:2:111361082:111495044:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR4435-2HG</td>\n",
       "      <td>GTGTGTACATCATTGGGAATGGAGGGAAATAAATGACTGGATGGTC...</td>\n",
       "      <td>0.417794</td>\n",
       "      <td>6654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113330</th>\n",
       "      <td>ENST00000575922.5</td>\n",
       "      <td>ENSG00000260005.6</td>\n",
       "      <td>chromosome:GRCh38:17:81228707:81233983:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>description:novel</td>\n",
       "      <td>ACATTTCTGAAGCCCTGCCACCTGCCATCCCTCGTAGGGATTCAAG...</td>\n",
       "      <td>0.647803</td>\n",
       "      <td>4506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19949</th>\n",
       "      <td>ENST00000623047.1</td>\n",
       "      <td>ENSG00000280145.4</td>\n",
       "      <td>chromosome:GRCh38:21:6630182:6638337:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>description:novel</td>\n",
       "      <td>ATGAGCCACTGTACTGCGCTGTGCCTACTTCAAAGGACTGAAAATA...</td>\n",
       "      <td>0.407179</td>\n",
       "      <td>5990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174648</th>\n",
       "      <td>ENST00000643616.1</td>\n",
       "      <td>ENSG00000229140.13</td>\n",
       "      <td>chromosome:GRCh38:8:128634199:128966001:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>CCDC26</td>\n",
       "      <td>GCGCTTCTTGCTGCCAACATGTGAAGAAAGACGTGTTTGCTTCCCC...</td>\n",
       "      <td>0.339852</td>\n",
       "      <td>20097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33903</th>\n",
       "      <td>ENST00000669227.1</td>\n",
       "      <td>ENSG00000235954.10</td>\n",
       "      <td>chromosome:GRCh38:22:27919484:28002588:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>TTC28-AS1</td>\n",
       "      <td>TTACGAGGGCTCCTCCCTTCAGCTTTGGTCCCTCGGGGCGTCTGGG...</td>\n",
       "      <td>0.556558</td>\n",
       "      <td>5322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id          ensembl_id  \\\n",
       "105314  ENST00000717492.1  ENSG00000245694.13   \n",
       "123700  ENST00000663612.1  ENSG00000224078.16   \n",
       "11323   ENST00000659277.1  ENSG00000224924.10   \n",
       "100329  ENST00000522627.1   ENSG00000253311.4   \n",
       "38960   ENST00000443641.5   ENSG00000234893.6   \n",
       "18874   ENST00000415033.4  ENSG00000235903.12   \n",
       "50616   ENST00000452075.7   ENSG00000293313.2   \n",
       "132712  ENST00000624299.1  ENSG00000247809.12   \n",
       "119139  ENST00000668181.1   ENSG00000259234.9   \n",
       "105431  ENST00000711040.1  ENSG00000250337.11   \n",
       "185086  ENST00000653715.1  ENSG00000230461.11   \n",
       "67677   ENST00000484322.5   ENSG00000242593.9   \n",
       "106417  ENST00000504820.2   ENSG00000249494.7   \n",
       "48835   ENST00000650366.1  ENSG00000229807.14   \n",
       "150406  ENST00000618132.2  ENSG00000251562.12   \n",
       "150518  ENST00000657283.1   ENSG00000254092.5   \n",
       "156796  ENST00000673654.1  ENSG00000172965.18   \n",
       "113330  ENST00000575922.5   ENSG00000260005.6   \n",
       "19949   ENST00000623047.1   ENSG00000280145.4   \n",
       "174648  ENST00000643616.1  ENSG00000229140.13   \n",
       "33903   ENST00000669227.1  ENSG00000235954.10   \n",
       "\n",
       "                                      chromosome_location gene_type  \\\n",
       "105314          chromosome:GRCh38:16:54905876:54930578:-1    lncRNA   \n",
       "123700           chromosome:GRCh38:15:25094037:25122467:1    lncRNA   \n",
       "11323           chromosome:GRCh38:21:20735784:20803297:-1    lncRNA   \n",
       "100329         chromosome:GRCh38:5:159776772:159871384:-1    lncRNA   \n",
       "38960   scaffold:GRCh38:HSCHR6_MHC_QBL_CTG1:1546099:15...    lncRNA   \n",
       "18874            chromosome:GRCh38:13:46052815:46116576:1    lncRNA   \n",
       "50616           chromosome:GRCh38:10:42513510:42552794:-1    lncRNA   \n",
       "132712          chromosome:GRCh38:15:96285068:96289265:-1    lncRNA   \n",
       "119139          chromosome:GRCh38:15:79187239:79283949:-1    lncRNA   \n",
       "105431            chromosome:GRCh38:5:27472227:27488861:1    lncRNA   \n",
       "185086         chromosome:GRCh38:1:213817996:213966287:-1    lncRNA   \n",
       "67677           chromosome:GRCh38:7:124032205:124395118:1    lncRNA   \n",
       "106417         chromosome:GRCh38:5:119007965:119070902:-1    lncRNA   \n",
       "48835            chromosome:GRCh38:X:73820684:73846270:-1    lncRNA   \n",
       "150406           chromosome:GRCh38:11:65501009:65506110:1    lncRNA   \n",
       "150518           chromosome:GRCh38:8:20950392:21112728:-1    lncRNA   \n",
       "156796         chromosome:GRCh38:2:111361082:111495044:-1    lncRNA   \n",
       "113330           chromosome:GRCh38:17:81228707:81233983:1    lncRNA   \n",
       "19949             chromosome:GRCh38:21:6630182:6638337:-1    lncRNA   \n",
       "174648         chromosome:GRCh38:8:128634199:128966001:-1    lncRNA   \n",
       "33903            chromosome:GRCh38:22:27919484:28002588:1    lncRNA   \n",
       "\n",
       "              gene_symbol                                           sequence  \\\n",
       "105314              CRNDE  AGACTCTCGAGCGCCGGCCCCAGGATGACAATCACATCCCAGGAGC...   \n",
       "123700             SNHG14  GGATCGATGATGACTTTTATACATGCATTCCTTGGAAAGCTGAACA...   \n",
       "11323           LINC00320  AATCATCTCCCATTGAAAGCTGAATCATGAGGATACCATCTCCAAG...   \n",
       "100329          LINC01847  GTTGCATTAAGTCATCGAGTTTGTGGTGCTTTGTTATGGGAGCCTT...   \n",
       "38960               HCG18  CAATACTCGCCGGAAGCGGAAGCCGCGCCGAGGCTCGTGTAAAAGT...   \n",
       "18874            CPB2-AS1  CTTCGTGGACACGCCCTCTAGGCGTAAGGTTCCATGCGAGCGGTAA...   \n",
       "50616             ZNF37BP  GTTTTTCTCCGGTACAGCCTGGTGAAGGATGCCCGGAGGCCTCGGC...   \n",
       "132712          NR2F2-AS1  TTTTCTAATAATTTTTTTTTTTTTTTTTTGAGACAGAGTCTCGCTC...   \n",
       "119139       ANKRD34C-AS1  AGTTAGAAGCAGATCAGAAATCGAACACGTTTGCTTTTGCGATTCG...   \n",
       "105431              PURPL  GGAGCCATGAGGAATTCATGCCTTGCAAAGGGGAGGAGCCTGCCCT...   \n",
       "185086          PROX1-AS1  CAGAGAGAACCATTCTCTGCTTTCAATAGGACAGAAGAGCTCCTTG...   \n",
       "67677   description:novel  GCCCCAGTACGTTGAAGAAGAAGGTGAAGAAGATCTCATCCAGAAA...   \n",
       "106417           DMXL1-DT  AATCTCTTTGAGGCTTCTACAGGAGAACGCCGTGCGTTTGAGCTTG...   \n",
       "48835                XIST  ACACAGCTCAACCTATCTGAATAAAATCCTACTCTCAGACCCCTTT...   \n",
       "150406             MALAT1  TTTTTTTTTTTTACACGAATTTGAGGAAAACCAAATGAATTTGATA...   \n",
       "150518  description:novel  TTCACCATGTTGGTCAGGCTGGTCTCGAATGCCTGACCTCAGGTGA...   \n",
       "156796        MIR4435-2HG  GTGTGTACATCATTGGGAATGGAGGGAAATAAATGACTGGATGGTC...   \n",
       "113330  description:novel  ACATTTCTGAAGCCCTGCCACCTGCCATCCCTCGTAGGGATTCAAG...   \n",
       "19949   description:novel  ATGAGCCACTGTACTGCGCTGTGCCTACTTCAAAGGACTGAAAATA...   \n",
       "174648             CCDC26  GCGCTTCTTGCTGCCAACATGTGAAGAAAGACGTGTTTGCTTCCCC...   \n",
       "33903           TTC28-AS1  TTACGAGGGCTCCTCCCTTCAGCTTTGGTCCCTCGGGGCGTCTGGG...   \n",
       "\n",
       "        gc_content  seq_length  \n",
       "105314    0.409809       19697  \n",
       "123700    0.388336        7339  \n",
       "11323     0.395783        4932  \n",
       "100329    0.414304        5607  \n",
       "38960     0.440120        4317  \n",
       "18874     0.392840        4078  \n",
       "50616     0.403728        8637  \n",
       "132712    0.445688        4198  \n",
       "119139    0.491772        5165  \n",
       "105431    0.342486        4625  \n",
       "185086    0.453307        5140  \n",
       "67677     0.357077        4310  \n",
       "106417    0.410514        4375  \n",
       "48835     0.399124       12555  \n",
       "150406    0.387116        4983  \n",
       "150518    0.402561        4685  \n",
       "156796    0.417794        6654  \n",
       "113330    0.647803        4506  \n",
       "19949     0.407179        5990  \n",
       "174648    0.339852       20097  \n",
       "33903     0.556558        5322  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract genes above 4000 nucleotides\n",
    "large_genes = gene_filtered[gene_filtered['seq_length'] > 4000]\n",
    "large_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d29d68eb-9def-4ce2-ae8e-29b5517383f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove ENST00000717492.1, ENST00000650366.1 and ENST00000643616.1 from dataframe(large, causes issues with RNAfold)\n",
    "transcripts_to_remove = ['ENST00000717492.1', 'ENST00000650366.1', 'ENST00000643616.1']\n",
    "gene_filtered = gene_filtered[~gene_filtered['transcript_id'].isin(transcripts_to_remove)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "82aea805-aa25-4a5e-bd2e-370df19f02c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize dictionaries to store structure and mfe\n",
    "structure_dict = {}\n",
    "mfe_dict = {}\n",
    "\n",
    "# Open and parse the RNAfold output file\n",
    "with open('/home/myoui/shared/ML/rnafold_output.txt', 'r') as f:\n",
    "    lines = f.readlines()\n",
    "    \n",
    "    for i in range(0, len(lines), 3):  # Read the file in chunks of 3 lines\n",
    "        # Extract transcript ID (remove '>' from the beginning)\n",
    "        transcript_id = lines[i].strip().replace('>', '')\n",
    "        \n",
    "        # Extract structure and MFE from line i+2\n",
    "        structure_line = lines[i+2].strip()\n",
    "        structure = structure_line.split(\" \")[0]\n",
    "        \n",
    "        # Check if MFE is present and can be converted to float\n",
    "        try:\n",
    "            mfe = float(structure_line.split(\" \")[1].strip(\"()\"))\n",
    "        except (IndexError, ValueError):\n",
    "            mfe = None  # Assign None if MFE is missing or can't be converted\n",
    "        \n",
    "        # Store structure and mfe in dictionaries\n",
    "        structure_dict[transcript_id] = structure\n",
    "        mfe_dict[transcript_id] = mfe\n",
    "\n",
    "# Add structure and mfe to gene_filtered dataframe by matching transcript_id\n",
    "gene_filtered['structure'] = gene_filtered['transcript_id'].map(structure_dict)\n",
    "gene_filtered['mfe'] = gene_filtered['transcript_id'].map(mfe_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8e2109e-a91d-4e6c-9e95-5d81e9600e0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_filtered['dot_count'] = gene_filtered['structure'].str.count(r'\\.')\n",
    "gene_filtered['bracket_count'] = gene_filtered['structure'].str.count(r'\\(') + gene_filtered['structure'].str.count(r'\\)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "6bd86adb-4089-42bb-b949-65f53bd0f4da",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = gene_filtered.dropna(subset=['mfe'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0e6d9d8d-2203-4cbe-b0a9-4751b56a2469",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4475/1951693723.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data['kmers'] = data['sequence'].apply(lambda x: getKmers(x))\n",
      "/tmp/ipykernel_4475/1951693723.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data['kmers'] = data['kmers'].apply(lambda x: ' '.join(x))\n"
     ]
    }
   ],
   "source": [
    "def getKmers(sequence, size=6):\n",
    "    return [sequence[x:x+size].lower() for x in range(len(sequence) - size + 1)]\n",
    "\n",
    "# Generate k-mers and add them as a feature\n",
    "data['kmers'] = data['sequence'].apply(lambda x: getKmers(x))\n",
    "\n",
    "# Combine k-mers into a single space-separated string per sequence\n",
    "data['kmers'] = data['kmers'].apply(lambda x: ' '.join(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "69002abb-63d7-4285-8b58-4bd5598dad14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>chromosome_location</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>sequence</th>\n",
       "      <th>gc_content</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>structure</th>\n",
       "      <th>mfe</th>\n",
       "      <th>dot_count</th>\n",
       "      <th>bracket_count</th>\n",
       "      <th>kmers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000707457.1</td>\n",
       "      <td>ENSG00000291414.1</td>\n",
       "      <td>scaffold:GRCh38:HG1369_PATCH:55354:55416:-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR4740</td>\n",
       "      <td>GCCAAGGACTGATCCTCTCGGGCAGGGAGTCAGAGGGGACCGCCCG...</td>\n",
       "      <td>0.682540</td>\n",
       "      <td>63</td>\n",
       "      <td>..((.((((.(((((((((((((.((...((....))..)))))))...</td>\n",
       "      <td>-37.4</td>\n",
       "      <td>17</td>\n",
       "      <td>46</td>\n",
       "      <td>gccaag ccaagg caagga aaggac aggact ggactg gact...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000707458.1</td>\n",
       "      <td>ENSG00000291415.1</td>\n",
       "      <td>scaffold:GRCh38:HG1369_PATCH:107309:107393:-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR3186</td>\n",
       "      <td>AGCCTGCGGTTCCAACAGGCGTCTGTCTACGTGGCTTCAACCAAGT...</td>\n",
       "      <td>0.564706</td>\n",
       "      <td>85</td>\n",
       "      <td>(((((..((((((((.((.(((((.(((.((((((((.(((...))...</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>17</td>\n",
       "      <td>68</td>\n",
       "      <td>agcctg gcctgc cctgcg ctgcgg tgcggt gcggtt cggt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000637028.1</td>\n",
       "      <td>ENSG00000283344.1</td>\n",
       "      <td>scaffold:GRCh38:HG1362_PATCH:80950:81034:1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1244-4</td>\n",
       "      <td>ATCTTATTCCGAGCATTCCAGTAACTTTTTTGTGTATGTACTTAGC...</td>\n",
       "      <td>0.317647</td>\n",
       "      <td>85</td>\n",
       "      <td>(((((((.....(((..((((..(((......((((.((((........</td>\n",
       "      <td>-15.0</td>\n",
       "      <td>35</td>\n",
       "      <td>50</td>\n",
       "      <td>atctta tcttat cttatt ttattc tattcc attccg ttcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000711167.1</td>\n",
       "      <td>ENSG00000292346.1</td>\n",
       "      <td>chromosome:GRCh38:Y:2609191:2609254:1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6089</td>\n",
       "      <td>CCCCGGGCCCGGCGTTCCCTCCCCTTCCGTGCGCCAGTGGAGGCCG...</td>\n",
       "      <td>0.843750</td>\n",
       "      <td>64</td>\n",
       "      <td>(((((..((((.(.(.((((..(((((..((...))..)))))..)...</td>\n",
       "      <td>-34.6</td>\n",
       "      <td>20</td>\n",
       "      <td>44</td>\n",
       "      <td>ccccgg cccggg ccgggc cgggcc gggccc ggcccg gccc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000711140.1</td>\n",
       "      <td>ENSG00000292355.1</td>\n",
       "      <td>chromosome:GRCh38:Y:1293918:1293992:1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR3690</td>\n",
       "      <td>CCCATCTCCACCTGGACCCAGCGTAGACAAAGAGGTGTTTCTACTC...</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>75</td>\n",
       "      <td>(((((((.(((.(((..((((.(((((....(((.........)))...</td>\n",
       "      <td>-32.8</td>\n",
       "      <td>25</td>\n",
       "      <td>50</td>\n",
       "      <td>cccatc ccatct catctc atctcc tctcca ctccac tcca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185318</th>\n",
       "      <td>ENST00000662321.1</td>\n",
       "      <td>ENSG00000177788.7</td>\n",
       "      <td>chromosome:GRCh38:1:229267157:229271051:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RAB4A-AS1</td>\n",
       "      <td>GGGCGGGGCGCGTGAGGACTTCGGCGCGCGCCGGAAGCACGTGCGC...</td>\n",
       "      <td>0.435714</td>\n",
       "      <td>2520</td>\n",
       "      <td>.(((..((((((((..(....)..))))))))(((((((.(((.((...</td>\n",
       "      <td>-784.5</td>\n",
       "      <td>926</td>\n",
       "      <td>1594</td>\n",
       "      <td>gggcgg ggcggg gcgggg cggggc ggggcg gggcgc ggcg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119055</th>\n",
       "      <td>ENST00000689196.1</td>\n",
       "      <td>ENSG00000231028.12</td>\n",
       "      <td>chromosome:GRCh38:6:135497944:135528351:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>AHI1-DT</td>\n",
       "      <td>AGGGGGCTCCCGCCTCCCCGGCCACGCGAGCGGGAACAGTCTCCTC...</td>\n",
       "      <td>0.398104</td>\n",
       "      <td>1055</td>\n",
       "      <td>.((((((....))))))..(((((..((((.((((......))))....</td>\n",
       "      <td>-249.8</td>\n",
       "      <td>451</td>\n",
       "      <td>604</td>\n",
       "      <td>aggggg gggggc ggggct gggctc ggctcc gctccc ctcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154910</th>\n",
       "      <td>ENST00000820942.1</td>\n",
       "      <td>ENSG00000248538.11</td>\n",
       "      <td>chromosome:GRCh38:8:9151725:9202973:1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>PPP1R3B-DT</td>\n",
       "      <td>AGCGCGCCCGGAGTGATCCGGCTGCGGGCCAGGGTCTGGGAAGGGG...</td>\n",
       "      <td>0.499084</td>\n",
       "      <td>1092</td>\n",
       "      <td>.((.((((((((....)))))..))).))((((((((((((.((((...</td>\n",
       "      <td>-335.1</td>\n",
       "      <td>474</td>\n",
       "      <td>618</td>\n",
       "      <td>agcgcg gcgcgc cgcgcc gcgccc cgcccg gcccgg cccg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72254</th>\n",
       "      <td>ENST00000648393.1</td>\n",
       "      <td>ENSG00000285666.2</td>\n",
       "      <td>chromosome:GRCh38:7:76818377:76903041:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>description:novel</td>\n",
       "      <td>TTCTGTTTTTTTTTCTCTCCCACCCATCCTTATTGCCAACCACCCC...</td>\n",
       "      <td>0.412121</td>\n",
       "      <td>1155</td>\n",
       "      <td>((((((((.....(((((...((((.(((((.((.(((.......(...</td>\n",
       "      <td>-290.3</td>\n",
       "      <td>457</td>\n",
       "      <td>698</td>\n",
       "      <td>ttctgt tctgtt ctgttt tgtttt gttttt tttttt tttt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80892</th>\n",
       "      <td>ENST00000504957.2</td>\n",
       "      <td>ENSG00000251377.3</td>\n",
       "      <td>chromosome:GRCh38:4:152206987:152264490:-1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>description:novel</td>\n",
       "      <td>ACTTTCCAATGAAGAGAAAACCTTTCCCATCAAAAGAGAGAATGGG...</td>\n",
       "      <td>0.448914</td>\n",
       "      <td>1243</td>\n",
       "      <td>..(((((......).))))((((((((..((......))...((((...</td>\n",
       "      <td>-351.5</td>\n",
       "      <td>487</td>\n",
       "      <td>756</td>\n",
       "      <td>actttc ctttcc tttcca ttccaa tccaat ccaatg caat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9295 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            transcript_id          ensembl_id  \\\n",
       "0       ENST00000707457.1   ENSG00000291414.1   \n",
       "1       ENST00000707458.1   ENSG00000291415.1   \n",
       "2       ENST00000637028.1   ENSG00000283344.1   \n",
       "3       ENST00000711167.1   ENSG00000292346.1   \n",
       "4       ENST00000711140.1   ENSG00000292355.1   \n",
       "...                   ...                 ...   \n",
       "185318  ENST00000662321.1   ENSG00000177788.7   \n",
       "119055  ENST00000689196.1  ENSG00000231028.12   \n",
       "154910  ENST00000820942.1  ENSG00000248538.11   \n",
       "72254   ENST00000648393.1   ENSG00000285666.2   \n",
       "80892   ENST00000504957.2   ENSG00000251377.3   \n",
       "\n",
       "                                  chromosome_location gene_type  \\\n",
       "0         scaffold:GRCh38:HG1369_PATCH:55354:55416:-1     miRNA   \n",
       "1       scaffold:GRCh38:HG1369_PATCH:107309:107393:-1     miRNA   \n",
       "2          scaffold:GRCh38:HG1362_PATCH:80950:81034:1     miRNA   \n",
       "3               chromosome:GRCh38:Y:2609191:2609254:1     miRNA   \n",
       "4               chromosome:GRCh38:Y:1293918:1293992:1     miRNA   \n",
       "...                                               ...       ...   \n",
       "185318     chromosome:GRCh38:1:229267157:229271051:-1    lncRNA   \n",
       "119055      chromosome:GRCh38:6:135497944:135528351:1    lncRNA   \n",
       "154910          chromosome:GRCh38:8:9151725:9202973:1    lncRNA   \n",
       "72254        chromosome:GRCh38:7:76818377:76903041:-1    lncRNA   \n",
       "80892      chromosome:GRCh38:4:152206987:152264490:-1    lncRNA   \n",
       "\n",
       "              gene_symbol                                           sequence  \\\n",
       "0                 MIR4740  GCCAAGGACTGATCCTCTCGGGCAGGGAGTCAGAGGGGACCGCCCG...   \n",
       "1                 MIR3186  AGCCTGCGGTTCCAACAGGCGTCTGTCTACGTGGCTTCAACCAAGT...   \n",
       "2               MIR1244-4  ATCTTATTCCGAGCATTCCAGTAACTTTTTTGTGTATGTACTTAGC...   \n",
       "3                 MIR6089  CCCCGGGCCCGGCGTTCCCTCCCCTTCCGTGCGCCAGTGGAGGCCG...   \n",
       "4                 MIR3690  CCCATCTCCACCTGGACCCAGCGTAGACAAAGAGGTGTTTCTACTC...   \n",
       "...                   ...                                                ...   \n",
       "185318          RAB4A-AS1  GGGCGGGGCGCGTGAGGACTTCGGCGCGCGCCGGAAGCACGTGCGC...   \n",
       "119055            AHI1-DT  AGGGGGCTCCCGCCTCCCCGGCCACGCGAGCGGGAACAGTCTCCTC...   \n",
       "154910         PPP1R3B-DT  AGCGCGCCCGGAGTGATCCGGCTGCGGGCCAGGGTCTGGGAAGGGG...   \n",
       "72254   description:novel  TTCTGTTTTTTTTTCTCTCCCACCCATCCTTATTGCCAACCACCCC...   \n",
       "80892   description:novel  ACTTTCCAATGAAGAGAAAACCTTTCCCATCAAAAGAGAGAATGGG...   \n",
       "\n",
       "        gc_content  seq_length  \\\n",
       "0         0.682540          63   \n",
       "1         0.564706          85   \n",
       "2         0.317647          85   \n",
       "3         0.843750          64   \n",
       "4         0.533333          75   \n",
       "...            ...         ...   \n",
       "185318    0.435714        2520   \n",
       "119055    0.398104        1055   \n",
       "154910    0.499084        1092   \n",
       "72254     0.412121        1155   \n",
       "80892     0.448914        1243   \n",
       "\n",
       "                                                structure    mfe  dot_count  \\\n",
       "0       ..((.((((.(((((((((((((.((...((....))..)))))))...  -37.4         17   \n",
       "1       (((((..((((((((.((.(((((.(((.((((((((.(((...))...  -44.9         17   \n",
       "2       (((((((.....(((..((((..(((......((((.((((........  -15.0         35   \n",
       "3       (((((..((((.(.(.((((..(((((..((...))..)))))..)...  -34.6         20   \n",
       "4       (((((((.(((.(((..((((.(((((....(((.........)))...  -32.8         25   \n",
       "...                                                   ...    ...        ...   \n",
       "185318  .(((..((((((((..(....)..))))))))(((((((.(((.((... -784.5        926   \n",
       "119055  .((((((....))))))..(((((..((((.((((......)))).... -249.8        451   \n",
       "154910  .((.((((((((....)))))..))).))((((((((((((.((((... -335.1        474   \n",
       "72254   ((((((((.....(((((...((((.(((((.((.(((.......(... -290.3        457   \n",
       "80892   ..(((((......).))))((((((((..((......))...((((... -351.5        487   \n",
       "\n",
       "        bracket_count                                              kmers  \n",
       "0                  46  gccaag ccaagg caagga aaggac aggact ggactg gact...  \n",
       "1                  68  agcctg gcctgc cctgcg ctgcgg tgcggt gcggtt cggt...  \n",
       "2                  50  atctta tcttat cttatt ttattc tattcc attccg ttcc...  \n",
       "3                  44  ccccgg cccggg ccgggc cgggcc gggccc ggcccg gccc...  \n",
       "4                  50  cccatc ccatct catctc atctcc tctcca ctccac tcca...  \n",
       "...               ...                                                ...  \n",
       "185318           1594  gggcgg ggcggg gcgggg cggggc ggggcg gggcgc ggcg...  \n",
       "119055            604  aggggg gggggc ggggct gggctc ggctcc gctccc ctcc...  \n",
       "154910            618  agcgcg gcgcgc cgcgcc gcgccc cgcccg gcccgg cccg...  \n",
       "72254             698  ttctgt tctgtt ctgttt tgtttt gttttt tttttt tttt...  \n",
       "80892             756  actttc ctttcc tttcca ttccaa tccaat ccaatg caat...  \n",
       "\n",
       "[9295 rows x 13 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6cc6a4b5-eda0-4907-b143-aa7f5acde4ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('/home/myoui/shared/ML/gene_list_para.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "500c4fa7-5cb6-4df3-8cb4-d3341d178e7d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          transcript_id         ensembl_id  \\\n",
      "1967  ENST00000363120.1  ENSG00000199990.1   \n",
      "1968  ENST00000365241.1  ENSG00000202111.1   \n",
      "1969  ENST00000365645.1  ENSG00000202515.1   \n",
      "1970  ENST00000602301.2  ENSG00000270123.4   \n",
      "2048  ENST00000418539.2  ENSG00000236824.2   \n",
      "2049  ENST00000389680.2  ENSG00000211459.2   \n",
      "2050  ENST00000387347.2  ENSG00000210082.2   \n",
      "\n",
      "                             chromosome_location  gene_type gene_symbol  \\\n",
      "1967   chromosome:GRCh38:5:140711275:140711373:1  vault_RNA    VTRNA1-1   \n",
      "1968   chromosome:GRCh38:5:140718925:140719013:1  vault_RNA    VTRNA1-2   \n",
      "1969   chromosome:GRCh38:5:140726158:140726246:1  vault_RNA    VTRNA1-3   \n",
      "1970  chromosome:GRCh38:5:136080470:136080597:-1  vault_RNA    VTRNA2-1   \n",
      "2048     chromosome:GRCh38:2:47335315:47335514:1      scRNA      BCYRN1   \n",
      "2049             chromosome:GRCh38:MT:648:1601:1    Mt_rRNA     MT-RNR1   \n",
      "2050            chromosome:GRCh38:MT:1671:3229:1    Mt_rRNA     MT-RNR2   \n",
      "\n",
      "                                               sequence  \n",
      "1967  GGGCTGGCTTTAGCTCAGCGGTTACTTCGACAGTTCTTTAATTGAA...  \n",
      "1968  GGGCTGGCTTTAGCTCAGCGGTTACTTCGAGTACATTGTAACCACC...  \n",
      "1969  GGGCTGGCTTTAGCTCAGCGGTTACTTCGCGTGTCATCAAACCACC...  \n",
      "1970  GGGTCGGAGTTAGCTCAAGCGGTTACCTCCTCATGCCGGACTTTCT...  \n",
      "2048  GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCTCTCAGGGAGG...  \n",
      "2049  AATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACA...  \n",
      "2050  GCTAAACCTAGCCCCAAACCCACTCCACCTTACTACCAGACAACCT...  \n"
     ]
    }
   ],
   "source": [
    "df = gene_list\n",
    "gene_types_of_interest = ['scRNA', 'Mt_rRNA', 'vault_RNA']\n",
    "filtered_df = df[df['gene_type'].isin(gene_types_of_interest)]\n",
    "print(filtered_df)\n",
    "filtered_df.to_csv('/home/myoui/shared/ML/scRNA_Mt_vault.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aae372a-6c85-43ad-a1ef-29e568aeee5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b40e41f1-9d88-47d1-95bf-8afb82e4b01a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy_env",
   "language": "python",
   "name": "scanpy_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
